dial_queue.go

```diff
@@ -0,0 +1,273 @@
+	DialQueueMaxParallelism    = 20
+	DialQueueMaxIdle           = 5 * time.Second
+	DialQueueScalingMutePeriod = 1 * time.Second
+)
+
+type dialQueue struct {
+	ctx    context.Context
+	dialFn func(context.Context, peer.ID) error
+
+	nWorkers      int
+	scalingFactor float64
+
+	in  *queue.ChanQueue
+	out *queue.ChanQueue
+
+	waitingCh chan waitingCh
+	dieCh     chan struct{}
+	growCh    chan struct{}
+	shrinkCh  chan struct{}
+}
+
+type waitingCh struct {
+	ch chan<- peer.ID
+	ts time.Time
+}
+
+// newDialQueue returns an adaptive dial queue that spawns a dynamically sized set of goroutines to preemptively
+// stage dials for later handoff to the DHT protocol for RPC. It identifies backpressure on both ends (dial consumers
+// and dial producers), and takes compensating action by adjusting the worker pool.
+//
+// Why? Dialing is expensive. It's orders of magnitude slower than running an RPC on an already-established
+// connection, as it requires establishing a TCP connection, multistream handshake, crypto handshake, mux handshake,
+// and protocol negotiation.
+//
+// We start with DialQueueMinParallelism number of workers, and scale up and down based on demand and supply of
+// dialled peers.
+//
+// The following events trigger scaling:
+// - we scale up when we can't immediately return a successful dial to a new consumer.
+// - we scale down when we've been idle for a while waiting for new dial attempts.
+// - we scale down when we complete a dial and realise nobody was waiting for it.
+//
+// Dialler throttling (e.g. FD limit exceeded) is a concern, as we can easily spin up more workers to compensate, and
+// end up adding fuel to the fire. Since we have no deterministic way to detect this for now, we hard-limit concurrency
+// to DialQueueMaxParallelism.
+func newDialQueue(ctx context.Context, target string, in *queue.ChanQueue, dialFn func(context.Context, peer.ID) error, nConsumers int) *dialQueue {
+	sq := &dialQueue{
+		ctx:           ctx,
+		dialFn:        dialFn,
+		nWorkers:      DialQueueMinParallelism,
+		scalingFactor: 1.5,
+
+		in:  in,
+		out: queue.NewChanQueue(ctx, queue.NewXORDistancePQ(target)),
+
+		growCh:    make(chan struct{}, nConsumers),
+		shrinkCh:  make(chan struct{}, 1),
+		waitingCh: make(chan waitingCh, nConsumers),
+		dieCh:     make(chan struct{}, DialQueueMaxParallelism),
+	}
+	for i := 0; i < DialQueueMinParallelism; i++ {
+		go sq.worker()
+	}
+	go sq.control()
+	return sq
+}
+
+func (dq *dialQueue) control() {
+	var (
+		p              peer.ID
+		dialled        = dq.out.DeqChan
+		resp           waitingCh
+		waiting        <-chan waitingCh
+		lastScalingEvt = time.Now()
+	)
+	for {
+		// First process any backlog of dial jobs and waiters -- making progress is the priority.
+		// This block is copied below; couldn't find a more concise way of doing this.
+		select {
+		case <-dq.ctx.Done():
+			// close channels.
+			if resp.ch != nil {
+				close(resp.ch)
+			}
+			for w := range waiting {
```

Comment:
Will this ever finish?

Reply:
Oh my, I rushed it and that was sloppy of me.